### PR TITLE
Fixing Linter errors: undefined-variables<E0602>.

### PR DESCRIPTION
**Pull request by kevinclark19a**
## Description

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/osutil/iosxe.py [hunk 1]
```diff
@@ -58,7 +59,7 @@ def publish_hostname(self, hostname):
         Restart NetworkManager first before publishing hostname
         """
         shellutil.run("service NetworkManager restart")
-        super(RedhatOSUtil, self).publish_hostname(hostname) # pylint: disable=E0602
+        super(IosxeOSUtil, self).publish_hostname(hostname)
```
Comment (kevinclark19a) [inline azurelinuxagent/common/osutil/iosxe.py:62]: This seems like it was copy-pasted from Redhat. Anyone know if this is the case?

File: azurelinuxagent/common/osutil/clearlinux.py [hunk 2]
```diff
@@ -75,18 +76,14 @@ def conf_sshd(self, disable_password):
...
             passwd_file_path = conf.get_passwd_file_path()
-            try:
-                passwd_content = fileutil.read_file(passwd_file_path)
-                if not passwd_content:
-                    # Empty file is no better than no file
-                    raise FileNotFoundError # pylint: disable=undefined-variable
-            except FileNotFoundError: # pylint: disable=undefined-variable
+            passwd_content = fileutil.read_file(passwd_file_path)
+            if not passwd_content:
                 new_passwd = ["root:*LOCK*:14600::::::"]
```
Comment (kevinclark19a) [inline azurelinuxagent/common/osutil/clearlinux.py:81]: Ah. Looking at this again, I'm seeing that this isn't correct. I didn't realize that `fileutil.read_file` could also throw a `FileNotFoundError`... The problem here is that python2 does not have that builtin.

Should `read_file` throw an error that works with python2?

File: azurelinuxagent/common/osutil/iosxe.py [hunk 3]
```diff
@@ -79,13 +80,13 @@ def get_instance_id(self):
         If that is missing, then extracts from dmidecode
         If nothing works (for old VMs), return the empty string
         '''
-        if os.path.isfile(PRODUCT_ID_FILE): # pylint: disable=E0602
+        if os.path.isfile(PRODUCT_ID_FILE):
```
Comment (kevinclark19a) [inline azurelinuxagent/common/osutil/iosxe.py:83]: `os` and `PRODUCT_ID_FILE` both seem like they were straight-up just not imported into this module...

File: azurelinuxagent/common/osutil/iosxe.py [hunk 4]
```diff
@@ -79,13 +80,13 @@ def get_instance_id(self):
...
             try:
-                s = fileutil.read_file(PRODUCT_ID_FILE).strip() # pylint: disable=E0602,C0103
+                s = fileutil.read_file(PRODUCT_ID_FILE).strip() # pylint: disable=C0103
                 return self._correct_instance_id(s.strip())
             except IOError:
                 pass
-        rc, s = shellutil.run_get_output(DMIDECODE_CMD) # pylint: disable=E0602,C0103
-        if rc != 0 or UUID_PATTERN.match(s) is None: # pylint: disable=E0602
+        rc, s = shellutil.run_get_output(DMIDECODE_CMD) # pylint: disable=C0103
+        if rc != 0 or UUID_PATTERN.match(s) is None:
```
Comment (kevinclark19a) [inline azurelinuxagent/common/osutil/iosxe.py:90]: Same as above, this time with `DMIDECODE_CMD` and  `UUID_PATTERN`.

File: azurelinuxagent/common/osutil/openwrt.py [hunk 5]
```diff
@@ -126,7 +126,7 @@ def restart_ssh_service(self): # pylint: disable=R1710
         if os.path.exists("/etc/init.d/sshd"): # pylint: disable=R1705
             return shellutil.run("/etc/init.d/sshd restart", chk_err=True)
         else:
-            logger.warn("sshd service does not exists", username) # pylint: disable=E0602
+            logger.warn("sshd service does not exists")
```
Comment (kevinclark19a) [inline azurelinuxagent/common/osutil/openwrt.py:129]: This seems like this was just copy-pasted from a method with `username` in scope.

File: azurelinuxagent/common/osutil/iosxe.py (hunk 1 above)
Comment (pgombar) [reply azurelinuxagent/common/osutil/iosxe.py:62]: Not sure, most of these distro-specific modules that are not endorsed came from external contributors.

File: azurelinuxagent/common/future.py [hunk 6]
```diff
@@ -25,16 +25,38 @@
 
     bytebuffer = memoryview # pylint: disable=C0103
 
+    # We aren't using these imports in this file, but we want them to be available
```
Comment (pgombar) [inline azurelinuxagent/common/future.py:28]: Thanks for cleaning these up!

**Review by pgombar — DISMISSED** [review]

File: azurelinuxagent/common/future.py [hunk 7]
```diff
@@ -25,16 +25,38 @@
...
+    #           This is intentional, so that code that wants to use builtins we're
+    #           assigning new names to doesn't need to check python versions before
+    #           doing so.
+
+    # pylint: disable=undefined-variable,invalid-name,redefined-builtin
+
+    ustr = unicode # Rename Python2 unicode to ustr 
+    bytebuffer = buffer
+    range = xrange
+    int = long
```
Comment (larohra) [inline azurelinuxagent/common/future.py:57]: Are these all for Py2?

File: azurelinuxagent/common/future.py [hunk 8]
```diff
@@ -25,16 +25,38 @@
 
     bytebuffer = memoryview # pylint: disable=C0103
 
+    # We aren't using these imports in this file, but we want them to be available
+    # to import from this module in others.
+    # Additionally, python2 doesn't have this, so we need to disable import-error
+    # as well.
+    from builtins import int, range # pylint: disable=unused-import,import-error
```
Comment (larohra) [inline azurelinuxagent/common/future.py:32]: Wont this throw if py2 doesnt have these?

File: azurelinuxagent/common/osutil/iosxe.py (hunk 3 above)
Comment (larohra) [reply azurelinuxagent/common/osutil/iosxe.py:83]: lol whoever wrote the code (and reviewed it) didnt do a good job :p

File: azurelinuxagent/common/future.py (hunk 7 above)
Comment (kevinclark19a) [reply azurelinuxagent/common/future.py:57]: Yes, that's right.

File: azurelinuxagent/common/future.py (hunk 8 above)
Comment (kevinclark19a) [reply azurelinuxagent/common/future.py:32]: It's just off screen in the GitHub view, but we're in a `if sys.version_info[0] == 3:`, so this shouldn't be executed in a python2 env.

Comment (larohra) [reply azurelinuxagent/common/future.py:32]: Ahh got it! Thanks!

File: azurelinuxagent/common/osutil/clearlinux.py [hunk 9]
```diff
@@ -79,14 +80,14 @@ def del_root_password(self):
                 passwd_content = fileutil.read_file(passwd_file_path)
                 if not passwd_content:
                     # Empty file is no better than no file
-                    raise FileNotFoundError # pylint: disable=undefined-variable
-            except FileNotFoundError: # pylint: disable=undefined-variable
+                    raise IOError
+            except (IOError, OSError):
```
Comment (narrieta) [inline azurelinuxagent/common/osutil/clearlinux.py:84]: should we be checking that errno is ENOENT?

Comment (narrieta) [inline azurelinuxagent/common/osutil/clearlinux.py:84]: why OSError?

**Review by narrieta — COMMENTED** [review]
just a couple of questions

File: azurelinuxagent/common/osutil/clearlinux.py (hunk 9 above)
Comment (kevinclark19a) [reply azurelinuxagent/common/osutil/clearlinux.py:84]: I can look up and link the specifics, but it has to do with python2/3 compatibility.

Comment (kevinclark19a) [reply azurelinuxagent/common/osutil/clearlinux.py:84]: Sorry, I'm not quite sure what this is referring to.

Comment (kevinclark19a) [reply azurelinuxagent/common/osutil/clearlinux.py:84]: Found the [stackoverflow post I referenced](https://stackoverflow.com/questions/28633555/how-to-handle-filenotfounderror-when-try-except-ioerror-does-not-catch-it).

Basically, `FileNotFoundError` is a subclass of `OSError` and a python3 exclusive, whereas `IOError` (again, in python3) is just an alias for `OSError`; in python2, though, file operations throw `IOError`s (though, again, **not** `FileNotFoundError`s). `IOError` is therefore needed to catch in python2, and `OSError` is the correct error to catch for python3.

I suppose we could just use `IOError` (as it is the error thrown in python2 and an alias for the error thrown in python3), but I think it's better to be more explicit with the `OSError` we're actually catching in python3.

Comment (narrieta) [reply azurelinuxagent/common/osutil/clearlinux.py:84]: I meant the previous code was checking for a specific error (FileNotFoundError); to do the same in the new code we would need to check that IOError.errno == ENOENT

Comment (narrieta) [reply azurelinuxagent/common/osutil/clearlinux.py:84]: OK

Comment (kevinclark19a) [reply azurelinuxagent/common/osutil/clearlinux.py:84]: That makes sense. I'll make the change.

**Review by narrieta — DISMISSED** [review]

**Review by larohra — APPROVED** [review]

**Review by narrieta — APPROVED** [review]

**Pull request merged**